Sources/SwiftMemcache/MemcachedResponseDecoder.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-memcache-gsoc open source project
+//
+// Copyright (c) 2023 Apple Inc. and the swift-memcache-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-memcache-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+import NIOPosix
+
+struct MemcachedResponseDecoder: NIOSingleStepByteToMessageDecoder {
+    /// Responses look like:
+    ///
+    /// <RC> <datalen*> <flag1> <flag2> <...>\r\n
+    ///
+    /// Where <RC> is a 2 character return code. The number of flags returned are
+    /// based off of the flags supplied.
+    ///
+    /// <datalen> is only for responses with payloads, with the return code 'VA'.
+    ///
+    /// Flags are single character codes, ie 'q' or 'k' or 'I', which adjust the
+    /// behavior of the command. If a flag requests a response flag (ie 't' for TTL
+    /// remaining), it is returned in the same order as they were in the original
+    /// command, though this is not strict.
+    ///
+    /// Flags are single character codes, ie 'q' or 'k' or 'O', which adjust the
+    /// behavior of a command. Flags may contain token arguments, which come after the
+    /// flag and before the next space or newline, ie 'Oopaque' or 'Kuserkey'. Flags
+    /// can return new data or reflect information, in the same order they were
+    /// supplied in the request. Sending an 't' flag with a get for an item with 20
+    /// seconds of TTL remaining, would return 't20' in the response.
+    ///
+    /// All commands accept a tokens 'P' and 'L' which are completely ignored. The
+    /// arguments to 'P' and 'L' can be used as hints or path specifications to a
+    /// proxy or router inbetween a client and a memcached daemon. For example, a
+    /// client may prepend a "path" in the key itself: "mg /path/foo v" or in a proxy
+    /// token: "mg foo Lpath/ v" - the proxy may then optionally remove or forward the
+    /// token to a memcached daemon, which will ignore them.
+    ///
+    /// Syntax errors are handled the same as noted under 'Error strings' section
+    /// below.
+    ///
+    /// For usage examples beyond basic syntax, please see the wiki:
+    /// https://github.com/memcached/memcached/wiki/MetaCommands
+    typealias InboundOut = MemcachedResponse
+
+    func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
```

Comment:
This is almost right however you have to account for the fact that there are multiple responses inside the buffer. Furthermore, we are doing a bunch of duplicated work here if we get partial responses e.g. if we get everything except the `\r\n` we are going to re-parse the return code and flags every time. To avoid this we should introduce a little state machine to this decoder. Some rough code how this might look

```
    private enum NextStep: Hashable {
        /// The initial step.
        case returnCode
        /// Decode the data length, flags or check if we are the end
        case dataLengthOrFlag(MemcachedResponse.ReturnCode)
        /// Decode the next flag
        case decodeNextFlag(MemcachedResponse.ReturnCode , UInt64, [Flag])
        /// Decode end of line
        case decodeEndOfLine(MemcachedResponse.ReturnCode, UInt64?, [Flag])
       etc...
    }

    private enum NextDecodeAction {
        /// We need more bytes to decode the next step.
        case waitForMoreBytes
        /// We can continue decoding.
        case continueDecodeLoop
        /// We have decoded the next response and need to return it.
        case returnDecodedResponse(MemcachedResponse)
    }

    /// The next step in decoding.
    private var nextStep: NextStep = .decodeReturnCode

    public mutating func decode(buffer: inout ByteBuffer) throws -> MemcachedResponse? {
        while self.nextStep != .none {
            switch try self.next(buffer: &buffer) {
            case .returnDecodedResponse(let response):
                return response

            case .waitForMoreBytes:
                return nil

            case .continueDecodeLoop:
                ()
            }
        }
        return nil
    }

    private mutating func next(buffer: inout ByteBuffer) throws -> NextDecodeAction {
        switch self.nextStep {
        case .decodeReturnCode:
           // Try to decode the return code and move the state
        case .decodeDataLengthOrFlag(let returnCode):
          // Check if the return code is `VA` then decode the data length otherwise move state to flag decoding
        }
    }

    public mutating func decodeLast(buffer: inout ByteBuffer, seenEOF: Bool) throws -> MemcachedResponse? {
        // Try to decode what is left in the buffer.
        if let output = try self.decode(buffer: &buffer) {
            return output
        }

        guard buffer.readableBytes == 0 || seenEOF else {
            // If there are still readable bytes left and we haven't seen an EOF
            // then something is wrong with the message or how we called the decoder.
            throw SomEError
        }

        switch self.nextStep {
           // We have to decide what to do here. Some steps are fine to be in.
        }
    }
```